Add sorting feauture

Leoism opened the pull request:
Allow posts to be sorted by new, top, and trending

Commit:
Add sorting feauture
Allow posts to be sorted by new, top, and trending

Commit:
Format SortServlet.java

Commit:
Swap usage of System time to use Clock

Commit:
Use Float.compare instead of if statements

dchin2 commented on capstone_project/src/main/java/com/google/sps/data/PostService.java:
```diff
@@ -107,6 +120,11 @@ public long upvotePost(HttpServletRequest request) {
 
     List<Entity> posts = datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
 
+    String sortType = request.getParameter("sort-type");
```
could this be null?

Leoism replied:
Technically it can, but the http request will always have a sort-type, unless a user modifies the html via the browser console.

dchin2 commented on capstone_project/src/main/java/com/google/sps/data/PostService.java:
```diff
@@ -115,4 +133,67 @@ public long upvotePost(HttpServletRequest request) {
+  }
+
+  /** Returns a Map of all comparators. */
+  private Map<String, Comparator<Entity>> initializeSorters() {
```
this can be static since it doesn't rely on any state of the class.

Leoism replied:
It can't be static. Although sortByNew and sortByTop don't rely on class variables, sortByTrending relies on getUpvoteRatio which relies on the clock instance of the class. This would cause initializeSorters to reply on the clock instance as well. I could make sortByNew and sortByTop static but it would look inconsistent, in my opinion

dchin2 commented on capstone_project/src/main/java/com/google/sps/data/PostService.java:
```diff
@@ -115,4 +133,67 @@ public long upvotePost(HttpServletRequest request) {
+  }
+
+  /** Returns a Comparator that sorts entities by new, highest timestamp to smallest timestamp. */
+  private Comparator<Entity> sortByNew() {
```
Here and other places could be static if it doesn't rely on the fields of the class and will be the same on each call.

lyndapham reviewed (approved):
lgtm! nice use of Float.compare(), very clean 👍 
@andrewnguyen @dchin2 for approval

Commit:
Add null check and tests for sort-type

Commit:
Improve Datastore retrieval
No longer rely on the location of the DOM element. Instead use the unique ID of the entity. This is different from the Key itself as is a unique String.

Leoism commented on capstone_project/src/test/java/com/google/sps/PostTest.java:
```diff
@@ -195,100 +209,4 @@ public void returnEmptyListOnInvalidSortMethod() {
     assertEquals(expected, actual);
   }
-
-  @Test
```
These tests were removed because the way upvotePost works has changed, causing these tests to no longer be relevant.

andrewnguyen commented on capstone_project/src/main/java/com/google/sps/data/PostService.java:
```diff
@@ -107,12 +120,86 @@ public long upvotePost(HttpServletRequest request) {
 
-    int postID = Integer.parseInt(request.getParameter("id"));
-    Entity postToUpvote = posts.get(postID);
+    long postID = Long.parseLong(request.getParameter("id"));
```
Can't you just retrieve the post by ID instead of fetching every post and scanning it for the right post?

andrewnguyen commented on capstone_project/src/main/java/com/google/sps/data/PostService.java:
```diff
@@ -107,12 +120,86 @@ public long upvotePost(HttpServletRequest request) {
+  /** Returns a Map of all comparators. */
+  private Map<String, Comparator<Entity>> initializeSorters() {
+    Map<String, Comparator<Entity>> sorters = new HashMap<>();
+    sorters.put("new", sortByNew());
```
My hunch is that you should move this map into a field and have them refer to method references https://docs.oracle.com/javase/tutorial/java/javaOO/methodreferences.html.

Commit:
Improve entity retrieval from datastore by recreating key

Commit:
Store method references into Map

andrewnguyen commented on capstone_project/src/main/java/com/google/sps/data/PostService.java:
```diff
@@ -103,16 +119,78 @@ public void storePost(HttpServletRequest request) {
+   * sort type does not exist, an empty list is returned.
+   */
+  public List<Entity> sortEntities(String sortType, List<Entity> entities) {
+    initializeSorters();
```
shouldn't initializeSorters() be in the constructor?

Leoism replied:
Yea. Yesterday I had it that way but it wasn't possible to mock the clock due to method references. The method references still referenced the original clock rather than the mocked clock, even though the class was referencing the mocked clock everywhere else (i think because the sorters were initialized before the mocked clock was set). I chose to go with a builder since it allows to set properties without requiring them in a constructor and method references can point to the mocked clock. This also let me move the initializeSorters method into the constructor

Commit:
Create a builder for PostService

dchin2 commented on capstone_project/src/main/java/com/google/sps/data/PostService.java:
```diff
@@ -103,16 +139,77 @@ public void storePost(HttpServletRequest request) {
+    postSorters.put("trending", this::sortByTrending);
+  }
+
+  /** Returns a Comparator that sorts entities by new, highest timestamp to smallest timestamp. */
```
Update comments,  this returns an int

dchin2 commented on capstone_project/src/main/java/com/google/sps/data/PostService.java:
```diff
@@ -103,16 +139,77 @@ public void storePost(HttpServletRequest request) {
+  }
+
+  /** Returns a comparator that sorts by trending, highest upvote ratio to lowest. */
+  private int sortByTrending(Entity first, Entity second) {
```
I don't quite understand why this and others can't be static? Could you explain a bit more.

Leoism replied:
The function relies on getUpvoteRatio which requires a clock instance. I cannot make non static method calls from a static method since there is no class instance, which means there is no clock instance. line 209 is where the clock instance is used.

Commit:
Update method comments for intializeSorters, sortByNew/Top/Trending
